### DISCUSSION TITLE
Web Server self health check

### DISCUSSION THREAD

**Discussion by czarop** [micropython/micropython#14384] [RP2040 / Pico]
Hello,

I am running a Microdot async webserver on a Pico W, listening on port 80 on the router. Mainly all good.
Problem is, on occasion, it silently stops responding to http requests. The web server 'serve' call does not return. This happens roughly every 1-3 days.
The Pico itself is alive and well, at least if I chuck another thread on there it stays active. Note - I am not using _thread, I just tested this out a couple of times.

Enter my naivete - bear with me:
Maybe this is a stupid idea, but I planned to get around this by having the pico send an http request to its own webserver, and restart if no response after a timeout.
This doesn't seem to work, at all, I cannot send a request to my own IP. I get an ECONNERROR 103 - ie cannot connect. Is this expected, and if so, could anyone explain why?
I am doing a simple urequest.post call. I have tried to use non-blocking requests too, including uaiohttpclient. Both working when sent from other devices.

Second, anyone have an idea of how I _can_ check for webserver responsiveness? or what might be causing the loss of responsiveness issue?
One approach would be a second pico W that can send request and reset the server pico via a wired connection... but i was hoping for something more elegant!

**Comment by peterhinch** [comment]
As a first step I would suggest a more sensitive test to how whether `asyncio` is still running.
```Python
from machine import Pin
task = None
async def toggle():
    led = Pin(''WL_GPIO0', Pin.OUT)
    while True:
        led(not led())
        await asyncio.sleep_ms(500)
```
Then in your Microdot code start the task, e.g.
```py
@app.route('/')
async def index(request):
    global task
    if task is None:
        task = asyncio.create_task(toggle())
    return 'Hello, world!'
```
If, after failure, the LED is still flashing, you know that your code and `asyncio` are still running. If the LED has stopped it is likely that the code has crashed. This probably has a hardware cause, which can be discussed and investigated. It would also mean that software-based recovery would not work: you would need to consider other approaches such as a watchdog timer.

Using `thread` did not really prove the same thing as the code on the main core might have locked up - the second core would continue, oblivious.

**Comment by czarop** [comment]
Thanks Peterhinch - logical suggestions. I have implemented this and we shall see what happens when it crashes again - lengthy testing process unfortunately! 
Even if this is hardware related (& I'd assume it's my code however!) it should be possible to cover all bases - a watchdog timer in case asyncio has frozen sounds good, and a periodic check since last http request received could work for the other senario, where the server has stopped functioning silently for whatever reason.
Currently the device receives a few guaranteed http request sper day, from another pico taking temp measurement from outside. So it could be forced to reboot after a while if these were not received. I'd like that to be rare - but if the code is running in the background then it's not a disaster to have the web server down for a few hours. Will see what the tests show & report back.

**Reply by peterhinch** [reply]
I have done a lot of long term testing when developing [mqtt_as](https://github.com/peterhinch/micropython-mqtt/tree/master/mqtt_as). It is possible to achieve long term running, but you need to be sure your power supply is of high quality. WiFi enabled boards can produce brief pulses of relatively high current, and Chinese wall-warts can be designed for charging phones rather than for long term running. The PSU's made for Raspberry Pi are very good.

I would also check that your application can survive a WiFi outage (I'm assuming here that your board is in station mode, connected to an AP/router).
